src/allmydata/web/common.py

```diff
@@ -466,31 +468,202 @@ def render(self, req):
 
 
-def render_exception(f):
+def _getChild_done(child, parent):
+    Message.log(
+        message_type=u"allmydata:web:common-getChild:result",
+        result=fullyQualifiedName(type(child)),
+    )
+    if child is None:
+        return resource.NoResource()
+    return child
+
+
+def _getChild_failed(reason):
+    text, code = humanize_failure(reason)
+    return resource.ErrorPage(code, "Error", text)
+
+
+def render_exception(render):
     """
     Decorate a ``render_*`` method with exception handling behavior to render
     an error page reflecting the exception.
     """
-    @wraps(f)
+    @wraps(render)
     def g(self, request):
-        try:
-            return f(self, request)
-        except Exception as e:
-            description, status = humanize_exception(e)
-            return resource.ErrorPage(status, "Error", description).render(request)
+        bound_render = render.__get__(self, type(self))
+
+        action = start_action(
+            action_type=u"allmydata:web:common-render",
+            uri=request.uri,
+            method=request.method,
+            handler=fullyQualifiedName(bound_render),
+        )
+        if getattr(request, "dont_apply_extra_processing", False):
+            with action:
+                return bound_render(request)
+
+        with action.context():
+            result = DeferredContext(maybeDeferred(bound_render, request))
+            result.addBoth(_finish, bound_render, request)
+            result.addActionFinish()
+        return NOT_DONE_YET
+
     return g
+
+
+def _finish(result, render, request):
+    if isinstance(result, Failure):
+        Message.log(
+            message_type=u"allmydata:web:common-render:failure",
+            message=result.getErrorMessage(),
+        )
+        _finish(
+            _renderHTTP_exception(request, result),
+            render,
+            request,
+        )
+    elif IResource.providedBy(result):
+        # If result is also using @render_exception then we don't want to
+        # double-apply the logic.  This leads to an attempt to double-finish
+        # the request.  If it isn't using @render_exception then you should
+        # fix it so it is.
+        Message.log(
+            message_type=u"allmydata:web:common-render:resource",
+            resource=fullyQualifiedName(type(result)),
+        )
+        result.render(request)
+    elif isinstance(result, bytes):
+        Message.log(
+            message_type=u"allmydata:web:common-render:bytes",
+        )
+        request.write(result)
+        request.finish()
+    elif isinstance(result, URLPath):
+        Message.log(
+            message_type=u"allmydata:web:common-render:URLPath",
+        )
+        if result.netloc == b"":
+            root = URLPath.fromRequest(request)
+            result.scheme = root.scheme
+            result.netloc = root.netloc
+        _finish(redirectTo(str(result), request), render, request)
+    elif isinstance(result, DecodedURL):
```

Comment:
It would probably be nice to get rid of the URLPath support and just use DecodedURL everywhere.